File: app/src/main/java/org/cuieney/videolife/ui/adapter/EssayAdapter.java
package org.cuieney.videolife.ui.adapter;

import android.content.Context;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import org.cuieney.videolife.R;
import org.cuieney.videolife.common.base.BaseRecycerViewAdapter;
import org.cuieney.videolife.common.image.ImageLoader;
import org.cuieney.videolife.common.utils.DateUtil;
import org.cuieney.videolife.entity.MusicListBean;
import org.cuieney.videolife.entity.YiRenBean.DataBean;

import java.util.List;

/**
 * Created by cuieney on 2017/5/17.
 */

public class EssayAdapter extends BaseRecycerViewAdapter<DataBean, RecyclerView.ViewHolder> {
    public EssayAdapter(Context context, List<DataBean> list) {
        super(context, list);
    }

    @Override
    public RecyclerView.ViewHolder getCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == R.layout.essay_item) {
            return new EssayAdapter.EssayHolder(inflater.inflate(viewType, parent, false));
        } else {

            return new EssayAdapter.TopHolder(inflater.inflate(viewType, parent, false));
        }
    }

    @Override
    public void getBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof EssayAdapter.EssayHolder) {

            EssayAdapter.EssayHolder viewHoler = (EssayAdapter.EssayHolder) holder;
            DataBean dataBean = list.get(position);
            ImageLoader.loadAll(context, dataBean.getImg_url(), viewHoler.image);
            ViewCompat.setTransitionName(viewHoler.image, String.valueOf(position) + "_image");
            viewHoler.title.setText(dataBean.getTitle());
            if (dataBean.getAuthor() != null) {
                viewHoler.author.setText("文 / "+dataBean.getAuthor().getUser_name());
            }
            viewHoler.subhead.setText(dataBean.getForward());
            if (dataBean.getTag_list() != null) {
                viewHoler.type.setText(dataBean.getTag_list().size() == 0?"- 阅读 -":"- "+dataBean.getTag_list().get(0).getTitle()+" -");
            }
            viewHoler.postTime.setText(DateUtil.formatDate2String(dataBean.getPost_date()));

            viewHoler.itemView.setOnClickListener(v -> {
                if (mClickListener != null) {
                    mClickListener.onItemClick(position,v,holder);
                }
            });
        }

    }


    @Override
    public int getItemViewType(int position) {

        if (position == 0) {
            return R.layout.top_item;
        }
        return R.layout.essay_item;

    }

    public static class EssayHolder extends RecyclerView.ViewHolder {

        private final TextView title;
        private final TextView type;
        private final TextView subhead;
        private final TextView postTime;
        private final TextView author;
        public final ImageView image;

        public EssayHolder(View itemView) {
            super(itemView);
            title = ((TextView) itemView.findViewById(R.id.title));
            type = ((TextView) itemView.findViewById(R.id.type));
            subhead = ((TextView) itemView.findViewById(R.id.subhead));
            postTime = ((TextView) itemView.findViewById(R.id.post_time));
            author = ((TextView) itemView.findViewById(R.id.author));
            image = ((ImageView) itemView.findViewById(R.id.image));
        }
    }

    public static class TopHolder extends RecyclerView.ViewHolder {
        private TextView textView;

        public TopHolder(View itemView) {
            super(itemView);
            textView = ((TextView) itemView.findViewById(R.id.name));
        }
    }
}

File: app/src/main/java/org/cuieney/videolife/kotlin/ui/adapter/EssayAdapter.kt
package org.cuieney.videolife.kotlin.ui.adapter

import android.support.v4.view.ViewCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import org.cuieney.videolife.R
import org.cuieney.videolife.entity.YiRenBean.DataBean
import kotlinx.android.synthetic.main.essay_item.view.*
import org.cuieney.videolife.common.image.ImageLoader
import org.cuieney.videolife.common.utils.DateUtil
import org.cuieney.videolife.ui.adapter.EssayAdapter

/**
 * Created by cuieney on 2017/5/23.
 */
class EssayAdapter(item:ArrayList<DataBean>): RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    var items:ArrayList<DataBean>
    init {
        items = ArrayList()
        addNews(item)
    }
    override fun getItemCount(): Int = items.size

    fun addNews(news: List<DataBean>) {
        val initPosition = items.size
        items.addAll(news)
        notifyItemRangeChanged(initPosition, items.size )
    }

    fun clearAndAddNews(news: List<DataBean>) {
        items.clear()
        notifyDataSetChanged()

        items.addAll(news)
        notifyItemRangeInserted(0, items.size)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is EssayHomeHolder) {
            holder.bindData(items[position],position)
            holder.itemView.setOnClickListener {
                clickListener.onItemClick(position,it,holder)
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        if (viewType == R.layout.essay_item) {
            return EssayHomeHolder(LayoutInflater.from(parent.context).inflate(viewType, parent, false))
        } else {

            return TopHolder(LayoutInflater.from(parent.context).inflate(viewType, parent, false))
        }

    }

    override fun getItemViewType(position: Int): Int {

        if (position == 0) {
            return R.layout.top_item
        }
        return R.layout.essay_item

    }

    class EssayHomeHolder(itemView: View) : RecyclerView.ViewHolder(itemView){
        fun bindData(dataBean:DataBean,position: Int){
            ImageLoader.loadAll(itemView.context, dataBean.getImg_url(),itemView.image)
            ViewCompat.setTransitionName(itemView.image, position.toString() + "_image")
            itemView.title.text = dataBean.title
            var author = dataBean.author.user_name
            itemView.author.text = "文 / $author"
            itemView.subhead.text = dataBean.forward

            itemView.type.text = if(dataBean.getTag_list().size == 0) "- 阅读 -" else "- " + dataBean.getTag_list().get(0).getTitle() + " -"
            itemView.post_time.text = DateUtil.formatDate2String(dataBean.getPost_date())
        }
    }

    class TopHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val textView: TextView

        init {
            textView = itemView.findViewById(R.id.name) as TextView
        }
    }

    fun setOnClickListener(mClickListener :OnItemClickListener){
        clickListener = mClickListener
    }

    lateinit var clickListener :OnItemClickListener
    interface OnItemClickListener {
        fun onItemClick(position: Int, view: View, vh: RecyclerView.ViewHolder)
    }

}